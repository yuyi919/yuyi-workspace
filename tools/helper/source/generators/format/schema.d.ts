import { PackageBuilder } from "../../schematics/internal-nx-plugins-lerna/schema";

export interface FormatGeneratorSchema {
  project: string;
  builder: PackageBuilder;
}
