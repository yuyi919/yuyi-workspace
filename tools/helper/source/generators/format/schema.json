{
  "$schema": "http://json-schema.org/schema",
  "cli": "nx",
  "id": "Format",
  "title": "",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "alias": "p",
      "$default": {
        "$source": "projectName"
      },
      "x-prompt": "What is the name of the project for the generator?"
    },
    "builder": {
      "type": "string",
      "enum": [
        "auto",
        "tsdx",
        "tsc"
      ],
      "description": "构建使用的工具(不选则根据原始选项，如果原始不属于任何一种则中断操作)",
      "default": "auto"
    },
    "tags": {
      "type": "string",
      "description": "Add tags to the project (used for linting)",
      "alias": "t"
    }
  },
  "required": [
    "project"
  ]
}
