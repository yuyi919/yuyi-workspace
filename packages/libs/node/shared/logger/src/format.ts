export enum LogLevel {
  Error = "error",
  Success = "success",
  Failed = "failed",
  Warn = "warn",
  Warning = "warning",
  Info = "info",
  Hint = "hint",
  Verbose = "verbose",
  Debug = "debug",
}

export const logLevels = {
  [LogLevel.Error]: 0,
  [LogLevel.Success]: 1,
  [LogLevel.Failed]: 1,
  [LogLevel.Warn]: 2,
  [LogLevel.Warning]: 2,
  [LogLevel.Info]: 2,
  [LogLevel.Hint]: 3,
  [LogLevel.Verbose]: 3,
  [LogLevel.Debug]: 4,
};
