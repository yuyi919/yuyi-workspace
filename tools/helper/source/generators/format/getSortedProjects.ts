import { ProjectGraph } from "@nrwl/workspace/src/core/project-graph";
import { ProjectNode } from "../../executors/build/getBuildablePackageJson";
import { ProjectType } from "@nrwl/workspace";
import { groupBy } from "lodash";
import { getSubType, LibraryType, SortedLibraryType } from "./LibraryType";
import { ProjectConfig } from "../../common/ProjectConfig";

export function getSortedProjects(
  projects: Record<string, ProjectConfig>,
  graph: ProjectGraph
): Record<string, ProjectConfig> {
  return sortObjectKeysWith(projects, (key) => {
    const { files, ...node } = (graph.nodes[key] as ProjectNode)?.data;
    const tags = node.nxJsonSection?.tags || node.tags || [];
    const libtypeIndex = SortedLibraryType.findIndex(
      (type) =>
        tags.includes(type) ||
        getSubType(type).some(
          (subType) => tags.includes(subType) || tags.includes(type + "-" + subType)
        )
    );
    const type = SortedLibraryType[libtypeIndex];
    console.log(key, type, tags);
    const subLibTypeIndex = getSubType(type).findIndex(
      (subType: string) => tags.includes(subType) || tags.includes(type + "-" + subType)
    );

    const libtype = SortedLibraryType[libtypeIndex];
    const projectType =
      projectTypeSort[
        libtype === LibraryType.Internal ? ProjectType.Library : projects[key].projectType
      ];
    return [projectType, libtypeIndex, subLibTypeIndex];
  });
}
export function sortObjectKeysWith<O extends Record<string, any>, K extends keyof O, R extends number | string | (number | string)[]>(
  collection: O,
  walkerCast: (key: K) => R
): O {
  const sortedProjects = {} as O;
  const group = groupBy(Object.keys(collection), (key: K) => {
    const lib = walkerCast(key);
    return lib instanceof Array ? lib.join("|") : lib;
  }) as Record<string, string[]>;
  const sortedProjectsKeys = [];
  let comp = (a, b) => (a == b ? a.localeCompare(b) : a - b);
  for (const key of Object.keys(group).sort(comp)) {
    // console.log(key, group[key].sort());
    for (const name of group[key].sort()) {
      sortedProjectsKeys.push(name);
      sortedProjects[name as keyof O] = collection[name];
    }
  }
  console.log(group, sortedProjectsKeys);
  return sortedProjects;
}

export const projectTypeSort: Record<ProjectType, number> = [
  ProjectType.Application,
  ProjectType.Library,
].reduce((r, key, index) => ({ ...r, [key]: index }), {} as any);
