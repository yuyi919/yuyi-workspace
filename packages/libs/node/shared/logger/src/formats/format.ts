/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable no-control-regex */
import colors from "colors";
import type { Color } from "colors";
import type { Format, TransformableInfo } from "logform";
import path from "path";
import triple from "triple-beam";
import { inspect } from "util";
import * as winston from "winston";
import { LogLevel, logLevels } from "../format";
import { calleeStore } from "./calleeStore";
import { getMeta } from "./init";
import { Callee, DevConsoleFormatOptions } from "./types";
declare module "logform" {
  export interface TransformableInfo {
    processStrackPath?: (path: string) => string;
    context?: string;
    stack?: string;
  }
}
declare module "colors" {
  interface Color {
    brightMagenta: Color;
    brightCyan: Color;
  }
  export const brightCyan: Color;
  export const brightMagenta: Color;
  export const brightYellow: Color;
  export const brightBlack: Color;
  export const brightBlue: Color;
  export const brightRed: Color;
}

export const nestLikeColors = {
  [LogLevel.Hint]: colors.brightCyan,
  [LogLevel.Info]: (str: string) => str,
  [LogLevel.Success]: colors.green,
  [LogLevel.Failed]: colors.brightRed,
  [LogLevel.Warn]: colors.yellow,
  [LogLevel.Warning]: colors.yellow,
  [LogLevel.Error]: colors.red,
  [LogLevel.Verbose]: colors.brightBlue,
  [LogLevel.Debug]: colors.grey,
};

export class DevConsoleFormat {
  private static readonly reSpaces = /^\s+/;
  private static readonly reSpacesOrEmpty = /^(\s*)/;
  private static readonly reColor = /\x1B\[\d+m/;

  private static readonly chars = {
    singleLine: "▪",
    startLine: "┏",
    line: "┃",
    endLine: "┗",
  };

  public constructor(public name: string, private opts: DevConsoleFormatOptions = {}) {
    if (typeof this.opts.addLineSeparation === "undefined") {
      this.opts.addLineSeparation = true;
    }
    if (typeof this.opts.showTimestamps === "undefined") {
      this.opts.showTimestamps = false;
    }
    if (typeof this.opts.inspectOptions === "undefined") {
      this.opts.inspectOptions = {
        depth: Infinity,
        colors: true,
        maxArrayLength: Infinity,
        breakLength: 120,
        compact: Infinity,
      };
    }
  }

  private inspector(value: any, metaLines: string[], noColor = false): void {
    const options = this.opts.inspectOptions ? { ...this.opts.inspectOptions } : {};
    if (noColor === true) {
      options.colors = false;
    }
    const inspector = typeof value === "string" ? value : inspect(value, options);
    // if (value != null && typeof value === 'object') {
    // console.log(inspector.replace(/([a-z]+)\{/, "{"))
    // inspector = dump(JSON.parse(inspector.replace(/([a-z]+)\{/, "{")), {
    // quotingType: '"'
    // })
    // }
    inspector.split("\n").forEach((line) => {
      metaLines.push(line);
    });
  }
  private upperFirst(level: string) {
    const [, color, keyword] = /^(\x1B\[\d+m)([a-zA-Z]+)/i.exec(level) || [];
    const at = keyword && level.indexOf(keyword);
    if (at) return color + keyword.charAt(0).toUpperCase() + level.slice(at + 1);
    return level;
  }

  private getPadding(message?: string): string {
    // const [, , keyword] = /^(\x1B\[\d+m)([a-zA-Z]+)/i.exec(level) || [];
    // return (" ").repeat(this.opts.context.length + 2 + 1 + keyword.length);
    let padding = "";
    const matches = message && message.match(DevConsoleFormat.reSpaces);
    if (matches && matches.length > 0) {
      padding = matches[0];
    }
    return padding;
  }

  private getMs(info: TransformableInfo): string {
    let ms = "";
    if (info.ms) {
      ms = colors.italic(colors.dim(` ${info.ms}`));
    }
    return ms;
  }
  static getTimestamp(timestamp: number) {
    const date = new Date(timestamp);
    return date.toISOString().replace(/T.+$/, " ") + date.toTimeString().replace(/ .+$/, "");
  }
  private getTimestamp(info: TransformableInfo): string {
    let timestamp = "";
    if (info.timestamp) {
      timestamp = DevConsoleFormat.getTimestamp(info.timestamp);
    }
    return colors.italic.brightMagenta.dim(timestamp);
  }

  private getStackLines(info: TransformableInfo): string[] {
    const stackLines: string[] = [];

    if (info.stack) {
      const error = new Error();
      error.stack = info.stack;
      this.inspector(error, stackLines, true);
      if (info.processStrackPath) {
        stackLines.forEach((line, index) => {
          stackLines[index] = line.replace(/\((.+)\)$/, (str, substr) => {
            return `(${info.processStrackPath(substr)})`;
          });
        });
      }
    }

    return stackLines;
  }

  private getMetaLines(info: TransformableInfo) {
    const plainItems: string[] = [];
    const metaLines: string[] = [];
    const lines = getMeta(info); //[SPLAT as unknown as string];
    if (lines?.length) {
      for (const line of lines) {
        if (line instanceof Array) {
          const splat = this.opts.immediate ? [...line] : line;
          this.inspector(splat, metaLines);
        } else if (line instanceof Object) {
          const splat = this.opts.immediate ? { ...line } : line;
          this.inspector(splat, metaLines);
        } else {
          this.inspector(line, metaLines);
          // plainItems.push(line as string);
        }
      }
    }
    return {
      plainItems,
      metaLines,
    };
  }

  private getCallee(): Callee | undefined {
    const callee = calleeStore?.value;

    if (callee?.filePath) {
      if (!this.opts.basePath) {
        // By default remove anything before and including `src/`
        callee.filePath = callee.filePath.replace(/^.*\/src\//, "");
      } else {
        callee.filePath = path.relative(this.opts.basePath, callee.filePath);
      }
    }

    return callee;
  }

  private getColor(info: TransformableInfo) {
    const textColor = ((nestLikeColors[info[triple.LEVEL]] as typeof nestLikeColors.info) ||
      nestLikeColors.info) as unknown as Color;
    const headColor = textColor === nestLikeColors.info ? colors.brightBlue : textColor;
    return { textColor, headColor };
  }

  private getMessage(
    info: TransformableInfo,
    isStart: boolean,
    headColor: Color,
    textColor: Color
  ): string[] {
    // eslint-disable-next-line prefer-const
    let [message, ...multipleLines] = info.message.split("\n");

    isStart = isStart || multipleLines.length > 0;
    const chr = DevConsoleFormat.chars[isStart ? "startLine" : "singleLine"];

    const hasMessage = message.replace(DevConsoleFormat.reSpacesOrEmpty, "").length;
    if (!hasMessage) message += `${colors.dim(colors.italic("(no message)"))}`;
    const space = message.match(/^(\s*)/)?.[0] || "";
    const plainMessage = space ? message.replace(space, "") : message;
    const context = info.context || this.opts.context;
    message = `${headColor(
      `:${space}${colors.dim(chr)} ${
        context && context !== this.name ? colors.yellow(`[${context}] `) : ""
      }`
    )}${textColor(plainMessage)}`;

    return [`${info.level}${message}`, ...multipleLines];
  }

  private write(
    info: TransformableInfo,
    multipleMessages: string[],
    metaLines: string[],
    headColor: Color,
    textColor: Color,
    callee?: Callee,
    showTrack?: boolean
  ): void {
    const infoIndex = triple.MESSAGE as unknown as string;
    const pad = this.getPadding(info.message);
    const hasMetaLines = !!metaLines.length;
    this.writeMultipleMessages(
      multipleMessages,
      info,
      headColor,
      textColor,
      pad,
      !hasMetaLines && !showTrack
    );
    //
    // if (this.opts.showTimestamps) {
    // info[infoIndex] += `\n${colors.dim(info.level)}:${pad}${color}${colors.dim(
    // DevConsoleFormat.chars.line
    // )}${this.getTimestamp(info)}`;
    // }

    if (showTrack) {
      const functionName = callee.functionName ? `${callee.functionName}` : "";
      const point = [callee.lineNumber, callee.lineColNum]
        .filter((o) => typeof o === "string")
        .join(":");
      const msg = ` at ${functionName}(${callee.filePath}${point ? `:${point}` : ""})`;

      info[infoIndex] += `\n${headColor(colors.dim(info.level + ":"))}${pad}${headColor(
        `${colors.dim(
          hasMetaLines ? DevConsoleFormat.chars.line : DevConsoleFormat.chars.endLine
        )} ${colors.yellow(colors.dim("[#]"))}${textColor(colors.dim(colors.italic(msg)))}`
      )}`;
    }
    this.writeMetaLines(metaLines, info, headColor, textColor, pad);
    if (this.opts.addLineSeparation) {
      info[infoIndex] += "\n";
    }
  }

  private writeMultipleMessages(
    metaLines: string[],
    info: TransformableInfo,
    headColor: Color,
    textColor: colors.Color,
    pad: string,
    isEnd?: boolean
  ) {
    const infoIndex = triple.MESSAGE as unknown as string;
    for (let lineNumberIndex = 0; lineNumberIndex < metaLines.length; lineNumberIndex++) {
      const line = metaLines[lineNumberIndex];
      const lineNumber = colors.dim(
        `[${(lineNumberIndex + 1).toString().padStart(metaLines.length.toString().length, " ")}]`
      );
      let chr = DevConsoleFormat.chars.line;

      if (isEnd && lineNumberIndex === metaLines.length - 1) {
        chr = DevConsoleFormat.chars.endLine;
      }

      info[infoIndex] += `\n${headColor(`${colors.dim(info.level)}:${pad}${colors.dim(chr)} `)}`;
      info[infoIndex] += colors.dim(`${colors.yellow(lineNumber)} ${textColor(line)}`);
    }
  }

  private writeMetaLines(
    metaLines: string[],
    info: TransformableInfo,
    headColor: Color,
    textColor: colors.Color,
    pad: string
  ) {
    const infoIndex = triple.MESSAGE as unknown as string;
    for (let lineNumberIndex = 0; lineNumberIndex < metaLines.length; lineNumberIndex++) {
      const line = metaLines[lineNumberIndex];
      const lineNumber = colors.dim(
        `[${(lineNumberIndex + 1).toString().padStart(metaLines.length.toString().length, " ")}]`
      );
      let chr = DevConsoleFormat.chars.line;

      if (lineNumberIndex === metaLines.length - 1) {
        chr = DevConsoleFormat.chars.endLine;
      }

      info[infoIndex] += `\n${headColor(`${colors.dim(info.level)}:${pad}${colors.dim(chr)} `)}`;
      info[infoIndex] += `${colors.yellow(lineNumber)} ${line}`;
    }
  }

  public transform(info: TransformableInfo): TransformableInfo {
    const { headColor, textColor } = this.getColor(info);
    info.level = `${
      this.opts.showTimestamps ? this.getTimestamp(info) + colors.brightCyan(" - ") : ""
    }${colors.brightCyan(`[${this.name}]`)} ${headColor(this.upperFirst(info.level))}`;
    const stackLine = this.getStackLines(info);
    const index = triple.MESSAGE as unknown as string;
    const callee = this.getCallee();
    const meta = this.getMetaLines(info);
    const metaLines: string[] = meta.metaLines;
    const showTrack = callee && (info[triple.LEVEL] === "error" || this.opts.showTrack);
    const [message, ...multipleMessages] = this.getMessage(
      info,
      metaLines.length > 0 || stackLine.length > 0 || showTrack,
      headColor,
      textColor
    );
    info[index] = message;
    info[index] = [info[index], ...meta.plainItems].join(" ");
    info[index] += this.getMs(info);

    // console.log("transform", info);
    this.write(
      info,
      multipleMessages.concat(stackLine),
      metaLines,
      headColor,
      textColor,
      callee,
      showTrack
    );
    return info;
  }
}

export const format = (name: string, opts?: DevConsoleFormatOptions): SelfFormat => {
  return Object.assign(
    winston.format.combine(
      ...[
        winston.format.timestamp(),
        winston.format.ms(),
        winston.format.errors({ stack: true }),
        winston.format.splat(),
        winston.format.json(),
        winston.format.padLevels({
          levels: logLevels,
        }),
        new DevConsoleFormat(name, opts)
      ].filter(Boolean)
    ),
    { opts }
  );
};
export type SelfFormat = Format & { opts?: DevConsoleFormatOptions };
