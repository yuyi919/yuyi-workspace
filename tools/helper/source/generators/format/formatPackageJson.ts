import { readJson, Tree, writeJson } from "@nrwl/devkit";
import { sortObjectByKeys } from "@nrwl/workspace/src/utils/ast-utils";
import { defaultsDeep } from "lodash";
import { PackageJSON } from "../../common/packageJsonUtils";
import { sortObjectKeysWith } from "./getSortedProjects";

export async function updatePackageJson(
  host: Tree,
  jsonPath: string,
  callback: (json: PackageJSON) => PackageJSON
) {
  const keywords = [
    "name",
    "version",
    "description",
    "author",
    "keywords",
    "license",
    "private",
    "main",
    "module",
    "types",
    "sideEffect",
    "scripts",
    "bin",
    "dependencies",
    "devDependencies",
    "peerDependencies",
    "files",
  ] as (keyof PackageJSON)[];
  const workspaceJson: PackageJSON = await readJson(host, jsonPath);
  const packageJson = defaultsDeep(callback(workspaceJson), {
    description: "",
    author: "",
  }) as PackageJSON;
  if (packageJson.keywords) packageJson.keywords = packageJson.keywords.sort();
  if (packageJson.scripts) packageJson.scripts = sortObjectByKeys(packageJson.scripts);
  if (packageJson.dependencies)
    packageJson.dependencies = sortObjectByKeys(packageJson.dependencies);
  if (packageJson.devDependencies)
    packageJson.devDependencies = sortObjectByKeys(packageJson.devDependencies);
  if (packageJson.peerDependencies)
    packageJson.peerDependencies = sortObjectByKeys(packageJson.peerDependencies);
  writeJson(
    host,
    jsonPath,
    sortObjectKeysWith(packageJson, (key) => {
      const index = keywords.indexOf(key);
      return index > -1 ? index : key;
    })
  );
}
