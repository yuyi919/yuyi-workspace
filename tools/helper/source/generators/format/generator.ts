import { readProjectConfiguration, Tree, updateProjectConfiguration } from "@nrwl/devkit";
import { createProjectGraph } from "@nrwl/workspace/src/core/project-graph";
import { defaultsDeep } from "lodash";
import * as Path from "path";
import { ProjectNode } from "../../executors/build/getBuildablePackageJson";
import { formatDeps } from "../../executors/build/updateDeps";
import { generateTscFiles } from "../../schematics/internal-nx-plugins-lerna/addLibFiles";
import { PackageBuilder } from "../../schematics/internal-nx-plugins-lerna/schema";
import { updateProject } from "../library/updateProject";
import { formatFiles } from "./format-files";
import { updatePackageJson } from "./formatPackageJson";
import { FormatGeneratorSchema } from "./schema";

export default async function (host: Tree, options: FormatGeneratorSchema) {
  const project = Object.assign(
    { builder: "auto" as PackageBuilder },
    readProjectConfiguration(host, options.project)
  );

  const graph = createProjectGraph();
  const node = (graph.nodes[options.project] as ProjectNode)?.data;
  const builder =
    (options.builder && options.builder !== "auto" ? options.builder : project.builder) || "auto";

  const {
    dependencies: deps,
    packageJson,
    packageJsonPath,
  } = formatDeps({
    workspaceRoot: process.cwd(),
    projectDir: options.project,
    // 如果该依赖项不为内部包，收集依赖
    match: (node, parent, deep) => deep < 1, // || !node.data.tags?.includes('internal')
  });

  if (node && node.projectType === "library" && node.tags?.includes("lerna-package")) {
    console.log("builder:", builder);
    if (builder === "tsc") {
      const tsconfigReferences = deps
        .filter((i) => i.node.data.projectType === "library" && i.node.data.builder === "tsc")
        .map((dep) => {
          const file = dep.node.data.files.find((i) => /tsconfig\.json$/.test(i.file));
          return file && Path.relative(node.root, file.file).replace(/\\/g, "/");
        })
        .filter(Boolean);
      console.log("tsconfig references:", tsconfigReferences);
      generateTscFiles(host, {
        name: options.project,
        projectRoot: project.root,
        references: tsconfigReferences.map((path) => (/^\./.test(path) ? path : "./" + path)),
      });
    }
    if (project.targets.build) {
      const buildOptions = project.targets.build.options || {};
      // 标准化outputs
      const outputs = [
        buildOptions.outputPath
          ? "{options.outputPath}"
          : buildOptions.cwd
          ? "{options.cwd}"
          : project.root,
      ];
      updateProjectConfiguration(host, options.project, {
        ...project,
        tags: node.tags,
        //@ts-ignore
        builder,
        targets: {
          ...project.targets,
          build: {
            ...project.targets.build,
            outputs,
          },
        },
      });
    }
  }

  updatePackageJson(host, packageJsonPath, () => {
    return defaultsDeep({
      ...packageJson,
      scripts: {
        ...packageJson.scripts,
        build: "heft build --clean",
        dev: "heft build --watch",
        test: "heft test",
        "test:watch": "heft test --watch",
        "nx:format": `nx run ${options.project}:command:format`,
      },
    });
  });

  updateProject(host, {
    name: options.project,
    packageManager: "pnpm",
    projectRoot: project.root,
  });
  // await updateWorkspace(host, (workspaceJson) => {
  //   return {
  //     ...workspaceJson,
  //     projects: getSortedProjects(workspaceJson.projects, graph),
  //   };
  // });
  return await formatFiles(host, graph);
}
