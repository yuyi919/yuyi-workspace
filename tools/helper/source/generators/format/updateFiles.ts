import { Tree, normalizePath } from "@nrwl/devkit";
import { defaultsDeep } from "lodash";
import * as Path from "path";
import {
  formatDeps,
  generateTscFiles,
  TypedProjectGraph,
  updatePackageJson,
  updateProject,
  getLibraryFromGraph
} from "../shared";
import { readProjectConfigurationWithBuilder } from "./generator";
import { FormatGeneratorSchema } from "./schema";


export async function updateFiles(
  host: Tree,
  options: FormatGeneratorSchema,
  graph: TypedProjectGraph
) {
  const project = readProjectConfigurationWithBuilder(host, options.project, options.builder);
  const node = getLibraryFromGraph(graph, options.project)?.data;
  const { builder } = project;

  const {
    dependencies: deps,
    packageJson,
    packageJsonPath,
  } = formatDeps({
    workspaceRoot: process.cwd(),
    projectName: options.project,
    // 如果该依赖项不为内部包，收集依赖
    match: (node, parent, deep) => deep < 1,
  });
  if (node?.projectType === "library") {
    console.log("builder:", builder);
    if (builder === "tsc") {
      const tsconfigReferences = deps
        .filter((i) => i.node.type === "lib" && i.node.data.builder === "tsc")
        .map((dep) => {
          const file = dep.node.data.files.find((i) => /tsconfig\.json$/.test(i.file));
          return file && normalizePath(Path.relative(node.root, Path.dirname(file.file)));
        })
        .filter(Boolean);
      // console.log("tsconfig references:", tsconfigReferences);
      generateTscFiles(host, {
        name: options.project,
        projectRoot: project.root,
        references: tsconfigReferences.map((path) => (/^\./.test(path) ? path : "./" + path)),
      });
    }
    if (project.targets.build) {
      //   const buildOptions = project.targets.build.options || {};
      //   // 标准化outputs
      //   const outputs = [
      //     buildOptions.outputPath
      //       ? "{options.outputPath}"
      //       : buildOptions.cwd
      //       ? "{options.cwd}"
      //       : project.root,
      //   ];
      //   updateProjectConfiguration(host, options.project, {
      //     ...project,
      //     tags: node.tags,
      //     //@ts-ignore
      //     builder,
      //     targets: {
      //       ...project.targets,
      //       build: {
      //         ...project.targets.build,
      //         outputs,
      //       },
      //     },
      //   });
    }
  }

  updatePackageJson(host, packageJsonPath, () => {
    return defaultsDeep({
      ...packageJson,
      scripts: {
        ...packageJson.scripts,
        build: "heft build --clean",
        "build:watch": "heft build --watch",
        dev: "heft build --watch",
        test: "heft test",
        "test:watch": "heft test --watch"
      },
    });
  });

  updateProject(
    host,
    {
      name: options.project,
      packageManager: "pnpm",
      projectRoot: project.root,
      builder,
    },
    project
  );
}
