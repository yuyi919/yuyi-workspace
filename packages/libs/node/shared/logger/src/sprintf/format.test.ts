import { sprintf } from "./format";

describe("Basic Rendering", () => {
  it("Basic Rendering", () => {
    expect(sprintf(`测试: %s %1$#T %%%d 然后`, "你好", -100, 1, 3)).toBe(
      "测试: 你好 String %-100 然后 1 3"
    );
  });
});

describe("JS: Rendering", () => {
  test(`Rendering Boolean Values, 'y' and 'Y' conversions`, () => {
    expect(sprintf("%O", {})).toMatchInlineSnapshot(`"0"`);
    expect(sprintf("%+7d", -42)).toMatchInlineSnapshot(`"    -42"`);
    expect(sprintf("|%S|", 1, () => {})).toMatchInlineSnapshot(`"|1| () => { }"`);
    expect(sprintf("|%1$y|%2$Y|%1$#Y|%2$#y|%2$.1y|", 1, 0)).toBe("|true|FALSE|YES|no|f|");
    expect(sprintf("|%05.2Y|%-5.2y|", 1, 0)).toBe("|000TR|fa   |");
  });
  test(`Rendering JSON | 'J' conversion`, () => {
    const x = {
      a: [1, [2, 3, 4], 5, 6, 7],
      b: {
        c: {
          d: { e: "f" },
          g: "h",
          i: "j",
        },
        k: "l",
        m: "n",
        o: "p",
      },
      q: "r",
    };
    expect(sprintf("%1$J", x)).toMatchSnapshot();
    expect(sprintf("%1$#J", x)).toMatchSnapshot();
  });
  it(`JS typeof, 'T' conversion`, () => {
    expect(sprintf("%1$T %#T", 1)).toBe("number Number");
    expect(sprintf("%T %1$#T", "foo")).toBe("string String");
    expect(sprintf("%T %1$#T", [1, 2, 3])).toBe("object Array");
    expect(sprintf("%T %1$#T", null)).toBe("object Null");
    expect(sprintf("%T %1$#T", undefined)).toBe("undefined Undefined");
  });
  it(`JS valueOf, 'V' conversion`, () => {
    const toString = () => '"string"';
    const valueOf = () => 3;
    expect(sprintf("%1$d %1$s %1$V", { toString })).toBe('0 "string" "string"');
    expect(sprintf("%1$d %1$s %1$V", { valueOf })).toBe("3 [object Object] 3");
    expect(sprintf("%1$d %1$s %1$V", { valueOf, [Symbol.toStringTag]: "Fake" })).toBe(
      "3 [object Fake] 3"
    );
    expect(sprintf("%1$d %1$s %1$V", { valueOf, toString })).toBe('3 "string" 3');
  });
});
