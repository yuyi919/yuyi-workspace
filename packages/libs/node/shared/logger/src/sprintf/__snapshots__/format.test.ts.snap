// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JS: Rendering Rendering JSON | 'J' conversion 1`] = `"{\\"a\\":[1,[2,3,4],5,6,7],\\"b\\":{\\"c\\":{\\"d\\":{\\"e\\":\\"f\\"},\\"g\\":\\"h\\",\\"i\\":\\"j\\"},\\"k\\":\\"l\\",\\"m\\":\\"n\\",\\"o\\":\\"p\\"},\\"q\\":\\"r\\"}"`;

exports[`JS: Rendering Rendering JSON | 'J' conversion 2`] = `"{\\"a\\":[1,[2,3,4],5,6,7],\\"b\\":{\\"c\\":{\\"d\\":{\\"e\\":\\"f\\"},\\"g\\":\\"h\\",\\"i\\":\\"j\\"},\\"k\\":\\"l\\",\\"m\\":\\"n\\",\\"o\\":\\"p\\"},\\"q\\":\\"r\\"}"`;
